connmgr/banmanager.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package connmgr
+
+import (
+	"net"
+	"sync"
+
+	"github.com/decred/dcrd/peer/v3"
+)
+
+// BanManagerConfig is the configuration struct for the ban manager.
+type BanManagerConfig struct {
+	// DisableBanning represents the status of disabling banning of
+	// misbehaving peers.
+	DisableBanning bool
+
+	// BanThreshold represents the maximum allowed ban score before
+	// misbehaving peers are disconnecting and banned.
+	BanThreshold uint32
+
+	// BanPeer represents a banning callback for cleanup and status
+	// updates purposes on a peer ban.
+	BanPeer func(p *peer.Peer)
+
+	// MaxPeers indicates the maximum number of inbound and outbound
+	// peers allowed.
+	MaxPeers int
+
+	// Whitelist represents the whitelisted IPs of the server.
+	WhiteList []net.IPNet
+}
+
+// banMgrPeer extends a peer to maintain additional state maintained by the
+// ban manager.
+type banMgrPeer struct {
+	*peer.Peer
+
+	isWhitelisted bool
+}
+
+// BanManager represents a peer ban score tracking manager.
+type BanManager struct {
+	cfg       BanManagerConfig
+	peers     map[string]*banMgrPeer
+	banScores map[string]*DynamicBanScore
```

Comment:
I would expect this to live in `banMgrPeer` as `banScore DynamicBanScore` (notice it's also not a pointer in the struct which results in less heap objects) since the purpose of that struct is to house per-peer state which is what this is.